internal/mtls/mtls.go

```diff
@@ -0,0 +1,130 @@
+	if !ok || len(clientCertData) == 0 {
+		return nil, nil, nil, fmt.Errorf("%w: 'tls.crt' key (client certificate)", ErrCertDataMissingInSecret)
+	}
+	logger.Debug("Successfully retrieved tls.crt (client certificate) from secret.")
+
+	clientKeyData, ok = secret.Data["tls.key"]
+	if !ok || len(clientKeyData) == 0 {
+		return nil, nil, nil, fmt.Errorf("%w: 'tls.key' key (client key)", ErrCertDataMissingInSecret)
+	}
+	logger.Debug("Successfully retrieved tls.key (client key) from secret.")
+
+	return caData, clientCertData, clientKeyData, nil
+}
+
+// loadMTLSCredentialsFromData loads mTLS credentials from byte slice.
+func loadMTLSCredentialsFromData(logger *zap.Logger, aksManaged bool, caCertData, clientCertData, clientKeyData []byte) (credentials.TransportCredentials, error) {
+	// Validate PEM data
+	if !isValidPEM(caCertData) || !isValidPEM(clientCertData) || !isValidPEM(clientKeyData) {
+		return nil, errors.New("invalid PEM data: ensure proper headers and formatting")
+	}
+
+	// Load client certificate and key from byte slices
+	cert, err := tls.X509KeyPair(clientCertData, clientKeyData)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load client certificate and key from data: %w", err)
+	}
+	logger.Debug("Client certificate and key loaded from data.")
+
+	// Load CA certificate from byte slice
+	certPool := x509.NewCertPool()
+	if !certPool.AppendCertsFromPEM(caCertData) {
+		return nil, errors.New("failed to append CA certificate to pool from data")
+	}
+	logger.Debug("CA certificate appended to pool from data.")
+
+	tlsConfig := &tls.Config{
+		ServerName:         hubbleRelayExpectedServerName,
+		Certificates:       []tls.Certificate{cert},
+		RootCAs:            certPool,
+		InsecureSkipVerify: false,
+		MinVersion:         tls.VersionTLS12,
+	}
+
+	var cred credentials.TransportCredentials
+	if aksManaged {
+		logger.Debug("AKS Managed: Using experimental NewTLSWithALPNDisabled")
+		cred = exp_credentials.NewTLSWithALPNDisabled(tlsConfig)
+	}
+
+	return cred, nil
+}
+
+// FetchAndLoadMTLSCredentials encapsulates the logic for fetching and loading mTLS credentials.
+func FetchAndLoadMTLSCredentials(ctx context.Context, clientset kubernetes.Interface, logger *zap.Logger, secretName, namespace string, isAKSManaged bool) (credentials.TransportCredentials, error) {
+	// Fetch mTLS certificates
+	caData, clientCertData, clientKeyData, fetchErr := getHubbleMTLSCertificatesFromSecret(
+		ctx, clientset, logger, secretName, namespace,
+	)
+	if fetchErr != nil {
+		logger.Error("Failed to fetch mTLS certificates", zap.Error(fetchErr))
+		return nil, fmt.Errorf("failed to fetch mTLS certificates: %w", fetchErr)
+	}
+
+	// Load credentials
+	creds, loadErr := loadMTLSCredentialsFromData(logger, isAKSManaged, caData, clientCertData, clientKeyData)
+	if loadErr != nil {
+		logger.Error("Failed to load mTLS credentials", zap.Error(loadErr))
+		return nil, fmt.Errorf("failed to load mTLS credentials: %w", loadErr)
```

Comment:
Let the caller log the error. No need to log multiple times.
```suggestion
		return nil, fmt.Errorf("failed to load mTLS credentials: %w", loadErr)
```